bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -276,21 +284,23 @@ private HeaderCacheElement getCachedHeader() throws IOException {
    */
   @VisibleForTesting
   boolean doRefresh() {
-    boolean requiresRefresh = false;
     synchronized (isRefreshing) {
       if (!isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good) {
         isRefreshing.set(true);
-        requiresRefresh = true;
+      } else {
+        return false;
       }
     }
-    if (!requiresRefresh) {
-      return false;
-    }
-    HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
-    synchronized (isRefreshing) {
-      headerCache.set(cacheElement);
-      isRefreshing.set(false);
-      isRefreshing.notifyAll();
+    try {
+      HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
+      synchronized (isRefreshing) {
```

Comment:
Ah, ok I see.

So the point only one of these queued requests (coming through doRefresh via asyncRefresh) will actually make it through the first synchronized block and *actually* issue the request. 

But so moving back a step, in the case where the header is stale, we're still likely creating/enqueueing a big burst of doRefresh calls via asyncRefresh(), most of which are going to just run through and not actually do anything. 

Why not just change the implementation of asyncRefresh so that it runs the first part (the check for whether a refresh should be enqueued) synchronously, and then just use the executor to enqueue the actual real I/O work i.e. refreshCredentialsWithRetry(). 

The simple and benignly racy way is to just use isRefreshing as a hint, i.e. just check if isRefreshing.get() is true before the call to executor.execute. That would cut down on most of the executor calls. Another way to accomplish this would be to introduce another CacheState -- `Refreshing` -- and skip multiple calls to asyncRefresh() that way.